xdsl/dialects/qssa.py

```diff
@@ -0,0 +1,144 @@
+from abc import ABC
+
+from xdsl.dialects.builtin import IntegerAttr, IntegerType
+from xdsl.ir import Dialect, ParametrizedAttribute, SSAValue, TypeAttribute
+from xdsl.irdl import (
+    IRDLOperation,
+    VarOpResult,
+    irdl_attr_definition,
+    irdl_op_definition,
+    operand_def,
+    prop_def,
+    result_def,
+    var_result_def,
+)
+from xdsl.parser import Parser
+from xdsl.printer import Printer
+
+
+@irdl_attr_definition
+class QubitAttr(ParametrizedAttribute, TypeAttribute):
+    """
+    Type for a single qubit
+    """
+
+    name = "qssa.qubit"
+
+
+qubit = QubitAttr()
+
+
+class QubitBase(IRDLOperation, ABC):
+    pass
+
+
+@irdl_op_definition
+class QubitAllocOp(QubitBase):
+    name = "qssa.alloc"
+
+    qubits = prop_def(IntegerAttr)
+
+    res: VarOpResult = var_result_def(qubit)
```

Comment:
From the semantics, this reads like it allocates a series of single qubits, rather than an array of qubits as the op is described in the paper. You could consider parameterising `QubitAttr` similar to `IntegerAttr`, or maybe @AntonLydike could advise?

Reply:
What I was saying is that these two ops are both completely valid, just one has redundant information:


```
@irdl_op_definition
class QubitAllocOp(QubitBase):
    name = "qssa.alloc"

    qubits = prop_def(IntegerAttr)

    res: VarOpResult = var_result_def(qubit)
```

and 
```
@irdl_op_definition
class QubitAllocOp(QubitBase):
    name = "qssa.alloc"

    res: VarOpResult = var_result_def(qubit)
```

Simply because we know that `qbits == len(res)`. Therefore we don't need to carry this additional information at all. The resulting IR would look like this:


```
%qbit1, %qbit2 = qssa.alloc
%qbit3 = qssa.alloc
```

So you just specify how many qbits you allocate by the number of results.

Reply:
For reference: 3.3.1 Allocation. `qssa.alloc` allocates an array of qubits of fixed or variable length. `%q = qssa.alloc : qubit<10>`

Reply:
I think @alexarice does not talk about qbit arrays, but models each qbit as a single SSA value. But that's for him to answer :smile:

Reply:
I seem to remember that semantically, an array of qubits may be understood to mean something different as compared to n single qubits, but I may be mistaken here.

Reply:
This is not directly based on the qssa dialect from the paper and is currently representing each individual qubit as a separate ssa value. I'd be open to renaming this dialect if this will cause confusion in the future

Reply:
The rest of the discussion didn't load for me on github, but I believe dropping the parameter is not ideal as the parser does not have access to the number of results (at least from my testing)